### PR TITLE
Add Pytorch Tensor Parallel support for Qwen2, Qwen2Moe, Starcoder2

### PR DESCRIPTION
**Pull request by VladOS95-cyber**
# What does this PR do?

This PR uses the torch.distributed.tensor.parallel subpackage to implement Tensor Parallel for Qwen2, Qwen2Moe, Starcoder2.
Fix qkv states dims for Mistral

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Link: https://github.com/huggingface/transformers/issues/34789
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@ArthurZucker

### CODE REVIEW

**Comment by VladOS95-cyber** [comment]
Hey @ArthurZucker! This PR is ready for review, please, take a look.

**Comment by VladOS95-cyber** [comment]
I think I should Fix Starcoder2 as well

**Comment by ArthurZucker** [comment]
Thanks, sorry a bit slow this week, reviewing asap!

File: src/transformers/models/qwen2_moe/configuration_qwen2_moe.py [hunk 1]
```diff
@@ -150,6 +150,17 @@ class Qwen2MoeConfig(PretrainedConfig):
     model_type = "qwen2_moe"
     keys_to_ignore_at_inference = ["past_key_values"]
 
+    # Default tensor parallel plan for base model `Qwe2Moe`
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/configuration_qwen2_moe.py:153]:
```suggestion
    # Default tensor parallel plan for base model `Qwen2Moe`
```

**Review by ArthurZucker — APPROVED** [review]
Lovely! Thanks for fixing and bringing more support!

**Comment by ArthurZucker** [comment]
Thanks @VladOS95-cyber ˜

**Pull request merged**